File: android/app/src/main/java/io/tapiwa/tapride/util/Geo.kt
package io.tapiwa.tapride.util

import kotlin.math.*

/**
 * Geographic utility functions used throughout TapRide.
 */
object Geo {

    private const val EARTH_RADIUS_KM = 6371.0

    /**
     * Computes the great-circle distance between two GPS coordinates
     * using the Haversine formula.
     *
     * @param lat1 Latitude of the first point in decimal degrees.
     * @param lng1 Longitude of the first point in decimal degrees.
     * @param lat2 Latitude of the second point in decimal degrees.
     * @param lng2 Longitude of the second point in decimal degrees.
     * @return Distance in kilometres.
     */
    fun haversine(lat1: Double, lng1: Double, lat2: Double, lng2: Double): Double {
        val dLat = Math.toRadians(lat2 - lat1)
        val dLng = Math.toRadians(lng2 - lng1)
        val a = sin(dLat / 2).pow(2) +
                cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2)) *
                sin(dLng / 2).pow(2)
        val c = 2 * asin(sqrt(a))
        return EARTH_RADIUS_KM * c
    }

    /**
     * Computes a bounding box around a centre point for querying nearby drivers.
     *
     * @param lat       Centre latitude.
     * @param lng       Centre longitude.
     * @param radiusKm  Radius of the bounding box in kilometres.
     * @return [BoundingBox] with min/max lat/lng values.
     */
    fun boundingBox(lat: Double, lng: Double, radiusKm: Double): BoundingBox {
        val latDelta = radiusKm / EARTH_RADIUS_KM * (180 / Math.PI)
        val lngDelta = radiusKm / (EARTH_RADIUS_KM * cos(Math.toRadians(lat))) * (180 / Math.PI)
        return BoundingBox(
            minLat = lat - latDelta,
            maxLat = lat + latDelta,
            minLng = lng - lngDelta,
            maxLng = lng + lngDelta
        )
    }

    /**
     * Represents a rectangular geographic bounding box.
     *
     * @property minLat Minimum latitude.
     * @property maxLat Maximum latitude.
     * @property minLng Minimum longitude.
     * @property maxLng Maximum longitude.
     */
    data class BoundingBox(
        val minLat: Double,
        val maxLat: Double,
        val minLng: Double,
        val maxLng: Double
    )
}

File: web/src/lib/geo.ts
/**
 * Geographic utility functions: haversine distance, bounding box,
 * and geohash encode/decode helpers via ngeohash.
 */
import ngeohash from 'ngeohash';

const EARTH_RADIUS_KM = 6371;

/** Convert degrees to radians */
const toRad = (deg: number): number => (deg * Math.PI) / 180;

/**
 * Calculate the haversine distance in kilometres between two coordinates.
 */
export function haversineKm(
  lat1: number,
  lng1: number,
  lat2: number,
  lng2: number,
): number {
  const dLat = toRad(lat2 - lat1);
  const dLng = toRad(lng2 - lng1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLng / 2) ** 2;
  return 2 * EARTH_RADIUS_KM * Math.asin(Math.sqrt(a));
}

/** Bounding box delta in degrees (~2 km) */
const BOX_DELTA = 0.02;

/**
 * Returns min/max lat/lng for a ~2 km bounding box around a point.
 * Used to query nearby drivers from Appwrite with simple range queries.
 */
export function boundingBox(lat: number, lng: number): {
  minLat: number;
  maxLat: number;
  minLng: number;
  maxLng: number;
} {
  return {
    minLat: lat - BOX_DELTA,
    maxLat: lat + BOX_DELTA,
    minLng: lng - BOX_DELTA,
    maxLng: lng + BOX_DELTA,
  };
}

/**
 * Encode a lat/lng pair to a geohash string at the given precision.
 */
export function encodeGeohash(lat: number, lng: number, precision = 7): string {
  return ngeohash.encode(lat, lng, precision);
}

/**
 * Decode a geohash string back to a { latitude, longitude } object.
 */
export function decodeGeohash(hash: string): { latitude: number; longitude: number } {
  return ngeohash.decode(hash);
}
